services/deployment-agent/src/simcore_service_deployment_agent/git_url_watcher.py

```diff
@@ -74,48 +96,93 @@ class GitRepo:  # pylint: disable=too-many-instance-attributes, too-many-argumen
     repo_id: str
     repo_url: URL
     branch: str
+    tags: str
     username: str
     password: str
     paths: List[Path]
     pull_only_files: bool
     directory: str = ""
 
+async def _checkout_repository(repo: GitRepo, tag: str = None):
+    if repo.pull_only_files:
+        await _git_checkout_files(repo.directory, repo.paths, tag)
+    else:
+        await _git_checkout_repo(repo.directory, tag)
 
-async def _init_repositories(repos: List[GitRepo]):
+async def _update_repository(repo: GitRepo):
+    if repo.pull_only_files:
+        await _git_pull_files(repo.directory, repo.paths)
+    else:
+        await _git_pull(repo.directory)
+
+async def _init_repositories(repos: List[GitRepo]) -> Dict:
+    description = {}
     for repo in repos:
         repo.directory = tempfile.TemporaryDirectory().name
         await _git_clone_repo(repo.repo_url, repo.directory, repo.branch, repo.username, repo.password)
-        if repo.pull_only_files:
-            await _git_checkout_files(repo.directory, repo.paths)
-        else:
-            await _git_checkout_repo(repo.directory)
-
-
-async def _check_repositories(repos: List[GitRepo]) -> Tuple[bool, str]:
-    change_detected = False
-    changes = ""
+        latest_tag = await _git_get_latest_matching_tag(repo.directory, repo.tags) if repo.tags else None
+        await _checkout_repository(repo, latest_tag)
+        sha = await _git_get_current_sha(repo.directory)
+        description[repo.repo_id] = f"{repo.repo_id}:{repo.branch}:{latest_tag}:{sha}" if latest_tag \
+            else f"{repo.repo_id}:{repo.branch}:{sha}"
+    return description
+
+async def _update_repo_using_tags(repo: GitRepo) -> Optional[str]:
+    log.debug("checking %s using tags", repo.repo_id)
+    # check if current tag is the latest and greatest
+    current_tag = await _git_get_current_matching_tag(repo.directory, repo.tags)
+    latest_tag = await _git_get_latest_matching_tag(repo.directory, repo.tags)
+    # there should always be a tag
+    assert latest_tag
+    if current_tag == latest_tag:
+        return

```

Comment:
MINOR: log.debug this decision ... this will help us follow the flow?

Reply:
done